Source/Core/InputCommon/ControllerInterface/WGInput/WGInput.h

```diff
@@ -0,0 +1,225 @@
+// Copyright 2018 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#pragma warning(push)
+// Disable warning for virtual functions with no virtual destructor:
+#pragma warning(disable : 4265)
+#include <wrl.h>
+#pragma warning(pop)
+
+#include <windows.gaming.input.h>
+
+#include "InputCommon/ControllerInterface/ControllerInterface.h"
+
+namespace ciface
+{
+namespace WGInput
+{
+namespace WGI = ABI::Windows::Gaming::Input;
+using Microsoft::WRL::ComPtr;
+
+void Init();
+void DeInit();
+void PopulateDevices();
+
+class Device : public Core::Device
+{
+protected:
+  class Button : public Core::Device::Input
+  {
+  public:
+    explicit Button(const boolean* _button) : m_button(*_button) {}
+    ControlState GetState() const override;
+
+  private:
+    // boolean comes from Windows API (typedef of unsigned char)
+    const boolean& m_button;
+  };
+
+  // A button with one of the "labels" that WGI provides:
+  class NamedButton : public Button
+  {
+  public:
+    NamedButton(const boolean* _button, const char* _name) : Button(_button), m_name(_name) {}
+    std::string GetName() const override;
+
+  private:
+    const char* const m_name;
+  };
+
+  // A button with no label so we name it by its index:
+  class IndexedButton : public Button
+  {
+  public:
+    IndexedButton(const boolean* _button, u32 _index) : Button(_button), m_index(_index) {}
+    std::string GetName() const override;
+
+  private:
+    u32 m_index;
+  };
+
+  class Axis : public Core::Device::Input
+  {
+  public:
+    Axis(const double* _axis, double _base, double _range)
```

Comment:
Would it not make sense to use a const reference here to enforce the non-null restriction on the API boundary, rather than silently allowing potential null pointer dereferences? Ditto for other relevant cases here.

Reply:
I feel taking a pointer protects against accidental creation references to temporaries.
I've run into this when passing e.g. a float instead of a double.
`Axis` stores a reference to passed argument and I think a null pointer is easier to debug than a dangling reference.